File: src/main/java/sage/web/PostController.java
package sage.web;

import java.io.IOException;
import java.util.Collection;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import sage.domain.commons.DomainRuntimeException;
import sage.domain.service.*;
import sage.entity.Blog;
import sage.entity.Tweet;
import sage.util.Colls;
import sage.web.auth.Auth;

@RestController
@RequestMapping(value = "/post", method = RequestMethod.POST)
public class PostController {

  private final Logger logger = LoggerFactory.getLogger(getClass());
  @Autowired
  private TweetPostService tweetPostService;
  @Autowired
  private BlogPostService blogService;
  @Autowired
  private TopicService topicService;
  @Autowired
  private GroupService groupService;

  @RequestMapping("/tweet")
  public boolean tweet(
      @RequestParam String content,
      @RequestParam(value = "pictureRef[]", defaultValue = "") Collection<String> pictureRefs,
      @RequestParam(value = "tagIds[]", defaultValue = "") Collection<Long> tagIds) {
    Long uid = Auth.checkCuid();
    logger.info("Got picture: " + pictureRefs);
    String tail = String.join(" ", Colls.map(pictureRefs, ref -> "img://" + ref));
    content = content+" "+tail;
    Tweet tweet = tweetPostService.post(uid, content, tagIds);
    logger.info("post tweet {} success", tweet.getId());
    return true;
  }

  @RequestMapping("/forward")
  public boolean forward(
      @RequestParam String content,
      @RequestParam Long originId,
      @RequestParam(value = "removedIds[]", defaultValue = "") Collection<Long> removedIds) {
    Long uid = Auth.checkCuid();
    Tweet tweet = tweetPostService.forward(uid, content, originId, removedIds);
    logger.info("forward tweet {} success", tweet.getId());
    return true;
  }

  @RequestMapping("/blog")
  public String blog(
      @RequestParam String title,
      @RequestParam String content,
      @RequestParam(value = "tagIds[]", defaultValue = "") Collection<Long> tagIds,
      @RequestParam(defaultValue = "true") Boolean sharable,
      @RequestParam(required = false) Long groupId) {
    Long uid = Auth.checkCuid();
    Blog blog = blogService.post(uid, title, content, tagIds);
    if (sharable) {
      tweetPostService.share(uid, blog);
    }
    if (groupId != null) {
      long topicId = topicService.post(uid, blog, groupId).getId();
      return String.format("/topic/%d", topicId);
    }
    logger.info("post blog {} success", blog.getId());
    return "/blog/" + blog.getId();
  }

  @RequestMapping("/edit-blog/{blogId}")
  public String editBlog(
      @PathVariable Long blogId,
      @RequestParam String title,
      @RequestParam String content,
      @RequestParam(value = "tagIds[]", defaultValue = "") Collection<Long> tagIds) {
    Long uid = Auth.checkCuid();
    Blog blog = blogService.edit(uid, blogId, title, content, tagIds);
    return "/blog/" + blog.getId();
  }

  @RequestMapping("/comment")
  public boolean comment(
      @RequestParam String content,
      @RequestParam Long sourceId,
      @RequestParam(required = false) Long replyUserId) {
    Long uid = Auth.checkCuid();
    //TODO save reply info in the comment
    tweetPostService.comment(uid, content, sourceId, replyUserId);
    return true;
  }
}

File: src/main/kotlin/sage/web/api/PostController.kt
package sage.web.api

import org.slf4j.LoggerFactory
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import sage.domain.commons.RichElement
import sage.web.auth.Auth
import sage.web.context.BaseController

@RestController
@RequestMapping("/post", method = arrayOf(RequestMethod.POST))
class PostController : BaseController() {

  private val logger = LoggerFactory.getLogger(javaClass)

  @RequestMapping("/tweet")
  fun tweet(
      @RequestParam content: String,
      @RequestParam("pictureRef[]", defaultValue = "") pictureRefs: Collection<String>): Boolean {
    val uid = Auth.checkUid()
    val tagIds = tagIds()
    logger.info("Got picture: " + pictureRefs)
    val richElements = pictureRefs.map { RichElement("picture", it) }
    val tweet = tweetPostService.post(uid, content, richElements, tagIds)
    logger.info("post tweet {} success", tweet.id)
    return true
  }

  @RequestMapping("/forward")
  fun forward(
      @RequestParam content: String,
      @RequestParam originId: Long,
      @RequestParam("removedIds[]", defaultValue = "") removedIds: Collection<Long>): Boolean {
    val uid = Auth.checkUid()
    val tweet = tweetPostService.forward(uid, content, originId, removedIds)
    logger.info("forward tweet {} success", tweet.id)
    return true
  }

  @RequestMapping("/comment")
  fun comment(
      @RequestParam content: String,
      @RequestParam sourceId: Long,
      @RequestParam(required = false) replyUserId: Long?,
      @RequestParam(required = false) forward: Boolean?): Boolean {
    val uid = Auth.checkUid()
    //TODO save reply info in the comment
    tweetPostService.comment(uid, content, sourceId, replyUserId)
    if (forward ?: false) {
      forward(content, sourceId, emptyList())
    }
    return true
  }
}
